x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/ranges/range_editor.tsx

```diff
@@ -0,0 +1,502 @@
+                    <EuiText size="s">{ltPrependLabel}</EuiText>
+                  </EuiToolTip>
+                }
+                fullWidth={true}
+                compressed={true}
+                placeholder={TO_PLACEHOLDER}
+                isInvalid={isValidNumber(to) && !isRangeWithin(tempRange)}
+              />
+            </EuiFlexItem>
+          </EuiFlexGroup>
+        </EuiFormRow>
+        <EuiFormRow fullWidth>
+          <EuiFieldText
+            compressed
+            value={safeLabel}
+            prepend={i18n.translate('xpack.lens.indexPattern.ranges.customIntervalPopoverLabel', {
+              defaultMessage: 'Label',
+            })}
+            onChange={({ target }) => {
+              const newRange = {
+                ...tempRange,
+                label: target.value,
+              };
+              setTempRange(newRange);
+              saveRangeIfValid(newRange);
+            }}
+          />
+        </EuiFormRow>
+      </EuiForm>
+    </EuiPopover>
+  );
+};
+
+const AdvancedRangeEditor = ({
+  ranges,
+  setRanges,
+  onToggleEditor,
+}: {
+  ranges: RangeType[];
+  setRanges: (newRanges: RangeType[]) => void;
+  onToggleEditor: () => void;
+}) => {
+  // use a local state to store ids with range objects
+  const [localRanges, setLocalRanges] = useState<LocalRangeType[]>(() =>
+    ranges.map((range) => ({ ...range, id: generateId() }))
+  );
+  // we need to force the open state of the popover from the outside in some scenarios
+  // so we need an extra state here
+  const [isOpenByCreation, setIsOpenByCreation] = useState(false);
+  const lastIndex = localRanges.length - 1;
+
+  const updateRanges = useCallback(
+    (updatedRanges: LocalRangeType[]) => {
+      setRanges(updatedRanges.map(({ id, ...rest }) => ({ ...rest })));
+      setLocalRanges(updatedRanges);
+    },
+    [setRanges]
+  );
+
+  const updateRangesDebounced = React.useMemo(() => debounce(updateRanges, 256), [updateRanges]);
+
+  const addNewRange = () => {
+    updateRanges([
+      ...localRanges,
+      {
+        id: generateId(),
+        from: -Infinity,
+        to: Infinity,
+        label: '',
+      },
+    ]);
+  };
+
+  const onDragEnd = ({
+    source,
+    destination,
+  }: {
+    source?: DraggableLocation;
+    destination?: DraggableLocation;
+  }) => {
+    if (source && destination) {
+      const items = euiDragDropReorder(localRanges, source.index, destination.index);
+      updateRangesDebounced(items);
+    }
+  };
+
+  return (
+    <EuiFormRow
+      label="Intervals"
+      labelAppend={
+        <EuiButtonEmpty
+          iconType="controlsHorizontal"
+          color="danger"
+          onClick={onToggleEditor}
+          size="xs"
+        >
+          {i18n.translate('xpack.lens.indexPattern.ranges.customIntervalsRemoval', {
+            defaultMessage: 'Remove custom intervals',
+          })}
+        </EuiButtonEmpty>
+      }
+    >
+      <>
+        <EuiDragDropContext onDragEnd={onDragEnd}>
```

Comment:
There's something weird about the drag and drop behavior: during the whole drag interaction I can't see the thing that I'm dragging.